### PR TITLE
8328181: C2: assert(MaxVectorSize >= 32) failed: vector length should be >= 32

### PR DESCRIPTION
**Pull request by jatin-bhateja**
This bug fix patch tightens the predication check for small constant length clear array pattern and relaxes associated feature checks. Modified few comments for clarity.

Kindly review and approve.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328181](https://bugs.openjdk.org/browse/JDK-8328181): C2: assert(MaxVectorSize &gt;= 32) failed: vector length should be &gt;= 32 (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18464/head:pull/18464` \
`$ git checkout pull/18464`

Update a local copy of the PR: \
`$ git checkout pull/18464` \
`$ git pull https://git.openjdk.org/jdk.git pull/18464/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18464`

View PR using the GUI difftool: \
`$ git pr show -t 18464`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18464.diff">https://git.openjdk.org/jdk/pull/18464.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18464#issuecomment-2016757634)

### CODE REVIEW

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1755]: Please add comment to clarify condition. I am reading it as ClearArray will not be supported for NOT avx512 because we can have vector length 512 bits for not avx512.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: This is only pertinent to known sized clear arrays which are optimized for AVX-512 targets, we already have such a check as part of matcher predicate, so removing it.

**Review by vnkozlov — APPROVED** [review]
This looks good. You need second review.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by TobiHartmann** [comment]
This introduced a performance regression, see [JDK-8332487](https://bugs.openjdk.org/browse/JDK-8332487). @jatin-bhateja, could you please have a look? Thanks.

File: src/hotspot/cpu/x86/x86.ad
Comment (eme64) [inline src/hotspot/cpu/x86/x86.ad:1753]: This seems problematic, and may lead to the regression in https://bugs.openjdk.org/browse/JDK-8332487

On non-AVX512 platforms, this is now always `true` instead of always `false`. Probably this was not intended, and you thought this way going to be default `false`?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad:1753]: Hi @eme64 , pattern predicates are checking the target features explicitly, this was introduced along with JDK-8257772, but looked redundant.

Comment (eme64) [reply src/hotspot/cpu/x86/x86.ad:1753]: I don't understand what you are implying. Are you saying this is not the reason for the regression?

**Pull request closed without merging**
